Tensile/Source/lib/include/Tensile/PropertyMatching.hpp

```diff
@@ -116,45 +116,235 @@ namespace Tensile
             }
         };
 
+        /**
+         * Shared code between the generic DistanceMatchingTable and the specialization
+         * for the special Equality distance
+         */
         template <typename Key,
                   typename Object,
                   typename Value,
                   typename ReturnValue,
                   typename Distance>
-        class DistanceMatchingTable : public MatchingTable<Object, Value, ReturnValue>
+        class DistanceMatchingCommon : public MatchingTable<Object, Value, ReturnValue>
         {
         public:
             using Base       = MatchingTable<Object, Value, ReturnValue>;
             using Entry      = MatchingTableEntry<Key, Value>;
             using Transform  = typename Base::Transform;
             using Properties = typename Base::Properties;
 
-            DistanceMatchingTable(ReturnValue nullValue = ReturnValue())
+            DistanceMatchingCommon(ReturnValue nullValue = ReturnValue())
                 : nullValue(nullValue)
             {
             }
 
-            DistanceMatchingTable(Properties const& properties,
-                                  ReturnValue       nullValue = ReturnValue())
+            DistanceMatchingCommon(Properties const& properties,
+                                   ReturnValue       nullValue = ReturnValue())
                 : Base(properties)
                 , nullValue(nullValue)
             {
             }
 
-            DistanceMatchingTable(Distance const&   distance,
-                                  Properties const& properties,
-                                  ReturnValue       nullValue = ReturnValue())
+            DistanceMatchingCommon(Distance const&   distance,
+                                   Properties const& properties,
+                                   ReturnValue       nullValue = ReturnValue())
                 : Base(properties)
                 , nullValue(nullValue)
                 , distance(distance)
             {
             }
 
+            virtual std::tuple<ReturnValue, double> findBestKeyMatch(Key const& key,
+                                                                     Transform transform) const = 0;
+
+            virtual std::tuple<ReturnValue, double>
+                findBestMatch(Object const& object, Transform transform) const override
+            {
+                return findBestKeyMatch(keyForProblem(object), transform);
+            }
+
+            virtual ReturnValue findBestEvaluationSolution(Object const&   object,
```

Comment:
This interface was designed to provide a run time switch, using an environment variable, to enable the experimental TAM selection method.  For future work we need to understand how to enable a clean way to select at runtime between two or more solutions methods. Does the tooling here provide such a mechanism?

Reply:
I don't think this PR adds any functionality that will be useful for changing the selection at runtime, at least not directly; all the changes are for build-time steps. I tried to make sure the changes I made wouldn't affect the TAM method, but I still need to verify it still works as expected. I will do this before I merge.

I know @afanfa also expressed desire for switching parts of solution selection at runtime, so I agree it's worth coming up with a plan to do this nicely and in a way that doesn't significantly impact selection speed.